include/Gaffer/ShufflePlug.h

```diff
@@ -0,0 +1,116 @@
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef GAFFER_SHUFFLEPLUG_H
+#define GAFFER_SHUFFLEPLUG_H
+
+#include "Gaffer/StringPlug.h"
+#include "Gaffer/TypedPlug.h"
+
+#include "IECore/CompoundObject.h"
+
+namespace Gaffer
+{
+
+/// Represents a "shuffle" - an name change for an existing value,
+/// with options to delete the original source name and an enable flag.
+class GAFFER_API ShufflePlug : public ValuePlug
+{
+
+	public :
+
+		GAFFER_PLUG_DECLARE_TYPE( ShufflePlug, ShufflePlugTypeId, ValuePlug );
+
+		ShufflePlug(
+			const std::string &name = defaultName<ShufflePlug>(),
+			Direction direction=In,
+			const std::string &source = "",
+			const std::string &destination = "",
+			bool deleteSource = false,
+			bool enabled = true,
+			unsigned flags = Default
+		);
+
+		StringPlug *sourcePlug();
+		const StringPlug *sourcePlug() const;
+
+		BoolPlug *enabledPlug();
+		const BoolPlug *enabledPlug() const;
+
+		StringPlug *destinationPlug();
+		const StringPlug *destinationPlug() const;
+
+		BoolPlug *deleteSourcePlug();
+		const BoolPlug *deleteSourcePlug() const;
+
+		bool acceptsChild( const GraphComponent *potentialChild ) const override;
+		Gaffer::PlugPtr createCounterpart( const std::string &name, Direction direction ) const override;
+
+};
+
+typedef FilteredChildIterator<PlugPredicate<Plug::Invalid, ShufflePlug> > ShufflePlugIterator;
+
+IE_CORE_DECLAREPTR( ShufflePlug )
+
+/// Represents a collection of shuffles, and provides methods for applying them
+/// to the likely data structures.
+class GAFFER_API ShufflesPlug : public ValuePlug
+{
+
+	public :
+
+		GAFFER_PLUG_DECLARE_TYPE( ShufflesPlug, ShufflesPlugTypeId, ValuePlug );
+
+		ShufflesPlug( const std::string &name=defaultName<ShufflesPlug>(), Direction direction=In, unsigned flags=Default );
+
+		bool acceptsChild( const GraphComponent *potentialChild ) const override;
+		bool acceptsInput( const Plug *input ) const override;
+		PlugPtr createCounterpart( const std::string &name, Direction direction ) const override;
+
+		/// Shuffles the sources into the destination container. The container type must have a std::pair value_type
+		/// and string-compatible keys (eg std::string, IECore::InternedString). Note also that non-matching sources
+		/// will not be copied into the destination container, so it will often be desirable to pre-fill the
+		/// destination with a copy of the source container before calling shuffle().
```

Comment:
Oh, I thought that's what we'd discussed... maybe I got confused. Pre-copying to destinationContainer is unavoidable in `ShufflePrimitiveVariables`, and I thought we'd decided it was better for `ShuffleAttributes` as well (`result->members() = inputAttributes->members();`). It was also easier to write the `shuffle()` algo that way, at least when I have the plugs as the outer loop...

I agree that'd it be nicer to not require this though, so hopefully it'll become easier to copy over non-matched sources when I re-write the loops based on your other comments.

Reply:
It may be me that got confused - I do remember talking about that stuff, but not the conclusion. The more I think about it, the more I prefer your earlier suggestion of returning the new container via `T shuffle( const T & )`. As I understand it, RVO and move assignment should ensure that there's not excessive overhead from doing that, and it's the cleanest and easiest from a caller's perspective (nothing to document, no opportunity to initialise the output container wrongly).

Reply:
Addressed in 294e070